// .ignored/@privy-io/node/public-api/services/key-quorums.mjs
import { generateAuthorizationSignatures } from "../../lib/authorization.mjs";
import { KeyQuorums, } from "../../resources.mjs";
export class PrivyKeyQuorumsService extends KeyQuorums {
    constructor(privyApiClient, privyClient) {
        super(privyApiClient);
        this.privyClient = privyClient;
    }
    async update(keyQuorumId, { authorization_context: authorizationContext = {}, ...params }) {
        const authorizationSignaturesHeader = await generateAuthorizationSignatures(this.privyClient, {
            authorizationContext,
            input: {
                version: 1,
                method: 'PATCH',
                url: `${this._client.baseURL}/v1/key_quorums/${keyQuorumId}`,
                body: params,
                headers: {
                    'privy-app-id': this._client.appID,
                },
            },
        });
        const response = await this._update(keyQuorumId, {
            ...params,
            'privy-authorization-signature': authorizationSignaturesHeader.join(','),
        });
        return response;
    }
    async delete(keyQuorumId, { authorization_context: authorizationContext = {}, ...params }) {
        const authorizationSignaturesHeader = await generateAuthorizationSignatures(this.privyClient, {
            authorizationContext,
            input: {
                version: 1,
                method: 'DELETE',
                url: `${this._client.baseURL}/v1/key_quorums/${keyQuorumId}`,
                body: params,
                headers: {
                    'privy-app-id': this._client.appID,
                },
            },
        });
        const response = await this._delete(keyQuorumId, {
            ...params,
            'privy-authorization-signature': authorizationSignaturesHeader.join(','),
        });
        return response;
    }
}
