// @privy-io/node@0.2.0/public-api/services/key-quorums.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PrivyKeyQuorumsService = void 0;
const authorization_1 = require("../../lib/authorization.js");
const resources_1 = require("../../resources.js");
class PrivyKeyQuorumsService extends resources_1.KeyQuorums {
    constructor(privyApiClient, privyClient) {
        super(privyApiClient);
        this.privyClient = privyClient;
    }
    async update(keyQuorumId, { authorization_context: authorizationContext = {}, ...params }) {
        const authorizationSignaturesHeader = await (0, authorization_1.generateAuthorizationSignatures)(this.privyClient, {
            authorizationContext,
            input: {
                version: 1,
                method: 'PATCH',
                url: `${this._client.baseURL}/v1/key_quorums/${keyQuorumId}`,
                body: params,
                headers: {
                    'privy-app-id': this._client.appID,
                },
            },
        });
        const response = await this._update(keyQuorumId, {
            ...params,
            'privy-authorization-signature': authorizationSignaturesHeader.join(','),
        });
        return response;
    }
    async delete(keyQuorumId, { authorization_context: authorizationContext = {}, ...params }) {
        const authorizationSignaturesHeader = await (0, authorization_1.generateAuthorizationSignatures)(this.privyClient, {
            authorizationContext,
            input: {
                version: 1,
                method: 'DELETE',
                url: `${this._client.baseURL}/v1/key_quorums/${keyQuorumId}`,
                body: params,
                headers: {
                    'privy-app-id': this._client.appID,
                },
            },
        });
        const response = await this._delete(keyQuorumId, {
            ...params,
            'privy-authorization-signature': authorizationSignaturesHeader.join(','),
        });
        return response;
    }
}
exports.PrivyKeyQuorumsService = PrivyKeyQuorumsService;
